### PR TITLE
 Bug 1499058 - TabDisplayManager use chained/imperative operation queue

### PR DESCRIPTION
**Pull request by garvankeeley**
- remove all state management in TDM
- weakify all self in closures
- private mode toggling logic consolidated into TDM
- drag-to-reorder operations are cancelled on *any* TabManager tab event (add, remove, select); deciding to impose this restriction was critical to the design of this code; without this restriction, the code will be more complicated and risky.

Uncertainties:
- with text in the search field and toggling private mode, I kept the search text and the search will be performed in the current mode, alternatively, the mode change could clear the search text
- if I can get a list of edge cases that the original code was designed to handle I'll write the tests to ensure this code also handles that.

All tests are enabled and passing locally.

### CODE REVIEW

File: Client/Frontend/Browser/TabDisplayManager.swift [hunk 1]
```diff
@@ -68,7 +80,76 @@ class TabDisplayManager: NSObject {
 
         tabManager.addDelegate(self)
         self.tabObservers = registerFor(.didLoadFavicon, .didChangeURL, queue: .main)
-        self.tabStore = self.tabsToDisplay
+
+        tabsToDisplay.forEach {
+            self.dataStore.insert($0)
```
[the remark this answers is not in the document]
Comment (garvankeeley) [reply Client/Frontend/Browser/TabDisplayManager.swift]: Assigning to a WeakList<>? IIRC there was no overload for that, I could add one if needed. This does suggest the reader that this isn't just a normal array so I didn't mind that.

[the remark this answers is not in the document]
Comment (farhanpatel) [reply Client/Frontend/Browser/TabDisplayManager.swift]: Ah. Didnt notice that. fair enough

**Comment by farhanpatel** [comment]
Not seeing a crash anymore when spamming tab actions. However when opening a new tab the toptabs tray doesnt move to focus the selected tab. You can see this when you open more tabs than can fit on the screen.

File: Client/Frontend/Browser/TabDisplayManager.swift [hunk 2]
```diff
@@ -252,297 +309,102 @@ extension TabDisplayManager: UICollectionViewDropDelegate {
...
-            self.updateTabsFrom(self.oldTabs, to: self.tabsToDisplay)
+        if let (type, operation) = operations.popLast() {
+            performingChainedOperations = true
+            collectionView.performBatchUpdates({ [weak self] in
+                // Baseline animation speed is 1.0, which is too slow, this (odd) code sets it to 3x
+                self?.collectionView.forFirstBaselineLayout.layer.speed = 3.0
+                operation()
+                }, completion: { [weak self] (done) in
+                    self?.performingChainedOperations = false
+                    self?.tabDisplayCompletionDelegate?.completedAnimation(for: type)
```
[the remark this answers is not in the document]
Comment (garvankeeley) [reply Client/Frontend/Browser/TabDisplayManager.swift]: Agreed I could format this better. This recursive function is intended to follow standard recursive termination style of: 
```
func recurse() {
  if conditionToExit { return }
  doWork()
  recurse()
}
```
Let me make that clearer.... (patching now...)

**Comment by garvankeeley** [comment]
@farhanpatel is this comment useful https://github.com/mozilla-mobile/firefox-ios/pull/4323/commits/b4743eed12f0b7fd42a20d99eebab18d82114327

**Comment by farhanpatel** [comment]
The tab focus doesnt seem to be always working. If you spam the add tab button on iPad it doesnt end up selecting the last tab once all the new tabs are added.

**Comment by garvankeeley** [comment]
Oopsy, I changed the tab focus behaviour when the selected tab isn't in view. Fixing...
